src/installment-sale/lib/logic.ergo

```diff
@@ -17,70 +17,81 @@ namespace org.accordproject.installmentsale
+    enforce (contract.MIN_PAYMENT.doubleValue <= state.balance_remaining.doubleValue) and (state.next_payment_month < 23);
+    enforce (contract.MIN_PAYMENT.doubleValue <= request.amount.doubleValue); // Underpaying is forbidden
+    enforce (request.amount.doubleValue <= state.balance_remaining.doubleValue); // overpaying is forbidden. this is NOT checked statically.
+    enforce (contract.INITIAL_DUE.currencyCode = request.amount.currencyCode); // Checks that the installment payment and initial due has the same currency
 
-    let before_interest = state.balance_remaining - request.amount;
+    let before_interest = state.balance_remaining.doubleValue - request.amount.doubleValue;
     let balance = before_interest * (1.0 + contract.INTEREST_RATE/100.00);
-    let total_paid = state.total_paid + request.amount;
+    let total_paid = state.total_paid.doubleValue + request.amount.doubleValue;
 
     set state InstallmentSaleState{
       stateId: "#1",
       status: "WaitingForFirstDayOfNextMonth",
-      balance_remaining: balance,
-      total_paid: total_paid,
+      balance_remaining: MonetaryAmount{ doubleValue : balance, currencyCode : contract.INITIAL_DUE.currencyCode},
+      total_paid: MonetaryAmount{ doubleValue : total_paid, currencyCode : contract.INITIAL_DUE.currencyCode},
       next_payment_month: state.next_payment_month + 1
     };
     emit PaymentObligation{
 			contract: contract,
 			promisor: some(contract.BUYER),
 			promisee: some(contract.SELLER),
 			deadline: none,
-			amount: MonetaryAmount{ doubleValue : request.amount, currencyCode : "USD" },
+			amount: request.amount,
 			description: contract.BUYER.partyId ++ " should pay installment to " ++ contract.SELLER.partyId
 		};
     return Balance{
-      balance: balance,
-      total_paid: total_paid
+      balance: MonetaryAmount{ doubleValue : balance, currencyCode : contract.INITIAL_DUE.currencyCode },
+      total_paid: MonetaryAmount{ doubleValue : total_paid, currencyCode : contract.INITIAL_DUE.currencyCode },
     }
   }
+  
   clause PayLastInstallment(request : ClosingPayment) : Balance emits PaymentObligation {
-    enforce (request.amount = state.balance_remaining + contract.DUE_AT_CLOSING);
-    let balance = state.balance_remaining + contract.DUE_AT_CLOSING - request.amount;
-    let total_paid = state.total_paid + request.amount;
+    enforce (roundn(request.amount.doubleValue, 2) = roundn(state.balance_remaining.doubleValue + contract.DUE_AT_CLOSING.doubleValue, 2));
+
+    let balance =  MonetaryAmount{ doubleValue : 0.0, currencyCode : contract.INITIAL_DUE.currencyCode};
+    let total_paid = MonetaryAmount{ doubleValue : state.total_paid.doubleValue + request.amount.doubleValue, currencyCode : contract.INITIAL_DUE.currencyCode};
+
     set state InstallmentSaleState{
       stateId: "#1",
       status: "Fulfilled",
       balance_remaining: balance,
       total_paid: total_paid,
       next_payment_month: 0
     };
-    emit PaymentObligation{
```

Comment:
So `PaymentObligation` for the `PayLastInstallment` clause was deleted, because I thought it doesn't make sense to have a payment obligation after making the 'last installment', especially since the description states:
`contract.BUYER.partyId ++ " should pay installment to " ++ contract.SELLER.partyId` when there is no installment left.

If it is beneficial to emit a message nonetheless, maybe we should change the message to 'CompletedPaymentObligation' or something like that? :O

Reply:
> So `PaymentObligation` for the `PayLastInstallment` clause was deleted, because I thought it doesn't make sense to have a payment obligation after making the 'last installment', especially since the description states:
> `contract.BUYER.partyId ++ " should pay installment to " ++ contract.SELLER.partyId` when there is no installment left.
> 
> If it is beneficial to emit a message nonetheless, maybe we should change the message to 'CompletedPaymentObligation' or something like that? :O

That's interesting! I had not thought of it that way. I think this is something that we might want to explain in the accord project documentation in a better way, but the payment obligation is something that is emitted so an external service could act on the actual payment.

In other words, the request for payment is checked by the contract and emits a real-world payment obligation (which can then be handed over to an online payment system, or so someone could write a check).

So concretely, it means the payment request occurs _before_ the actual payment occurs, which may be confusing in this particular template, but is consistent with the current programming model.

I hope that makes sense.